File: rest-server/app/service/job.js
// Copyright (c) Microsoft Corporation
// All rights reserved.
//
// MIT License
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'use strict';

const Service = require('egg').Service;
const execa = require('execa');
const mustache = require('mustache');
const ipmaptable = require('../../util/ipMapTable');
const Hdfs = require('../../util/hdsf');
const fs = require('fs');
const LError = require('../error/proto');
const ECode = require('../error/code');
const yarnContainerScriptTemplate = require('../templates/yarnContainerScript');
const dockerContainerScriptTemplate = require('../templates/dockerContainerScript');

class TokenService extends Service {

  constructor(...args) {
    super(...args);
    this.launcherConfig = this.config.launcherConfig;
  }

  async _invokeService(url, options = { dataType: 'json' }) {
    const launcherConfig = this.launcherConfig;
    const requestRes = await this.ctx.curl(url,
      Object.assign({ headers: launcherConfig.webserviceRequestHeaders }, options)
    );
    // if (requestRes.status !== 200) {
    //   throw new LError(ECode.REMOTE_INVOKE_ERROR, requestRes.data);
    // }
    const requestResJson = typeof requestRes.data === 'object' ?
      requestRes.data : JSON.parse(requestRes.data);
    return requestResJson;
  }

  async checkMinTaskNumber(job) {
    if ('killAllOnCompletedTaskNumber' in job) {
      const errorMessage = 'killAllOnCompletedTaskNumber has been obsoleted, please use minFailedTaskCount and minSucceededTaskCount instead.';
      throw new LError(ECode.INVALID_PARAM, errorMessage);

    }
    for (let i = 0; i < job.taskRoles.length; i++) {
      const taskNumber = job.taskRoles[i].taskNumber;
      const minFailedTaskCount = job.taskRoles[i].minFailedTaskCount || 0;
      const minSucceededTaskCount = job.taskRoles[i].minSucceededTaskCount || 0;
      if (minFailedTaskCount > taskNumber || minSucceededTaskCount > taskNumber) {
        const errorMessage = 'minFailedTaskCount or minSucceededTaskCount should not be greater than tasks number.';
        throw new LError(ECode.INVALID_PARAM, errorMessage);
      }
    }
  }

  async getJob(jobName) {
    const launcherConfig = this.launcherConfig;
    const frameworkPath = launcherConfig.frameworkPath(jobName);
    const requestResJson = await this._invokeService(frameworkPath);
    if (!requestResJson) {
      throw new LError(ECode.NOT_FOUND, `Job ${jobName} is not found.`);
    }
    if (requestResJson.exception) {
      return;
    }
    const jobDetail = await this.generateJobDetail(requestResJson);
    return Object.assign({ name: jobName }, jobDetail);
  }

  async putJob(name, data) {
    const launcherConfig = this.launcherConfig;
    if (!data.originalData.outputDir) {
      data.outputDir = `${launcherConfig.hdfsUri}/Output/${data.userName}/${name}`;
    }
    for (const fsPath of [ 'authFile', 'dataDir', 'outputDir', 'codeDir' ]) {
      data[fsPath] = data[fsPath].replace('$PAI_DEFAULT_FS_URI', launcherConfig.hdfsUri);
      data[fsPath] = data[fsPath].replace(/\$PAI_JOB_NAME(?![\w\d])/g, name);
      data[fsPath] = data[fsPath].replace(/(\$PAI_USER_NAME|\$PAI_USERNAME)(?![\w\d])/g, data.userName);
    }
    await this.service.user.checkUserVc(data.userName, data.virtualCluster);

    await this._initializeJobContextRootFolders();
    await this._prepareJobContext(name, data);
    await this._invokeService(launcherConfig.frameworkPath(name), { method: 'PUT', data: this.generateFrameworkDescription(data) });

  }

  async _initializeJobContextRootFolders() {
    const launcherConfig = this.launcherConfig;
    const hdfs = new Hdfs(launcherConfig.webhdfsUri);
    const hdfsTasks = [
      hdfs.createFolder(
        '/Output',
        { 'user.name': 'root', permission: '777' }
      ),
      hdfs.createFolder(
        '/Container',
        { 'user.name': 'root', permission: '777' }
      ),
    ];
    await Promise.all(hdfsTasks);
  }

  async _prepareJobContext(name, data) {
    const launcherConfig = this.launcherConfig;
    const hdfs = new Hdfs(launcherConfig.webhdfsUri);
    const hdfsTasks = [
      hdfs.createFile(
        `/Container/${data.userName}/${name}/${launcherConfig.jobConfigFileName}`,
        JSON.stringify(data.originalData, null, 2),
        { 'user.name': data.userName, permission: '644', overwrite: 'true' }
      ),
      hdfs.createFile(
        `/Container/${data.userName}/${name}/${launcherConfig.frameworkDescriptionFilename}`,
        JSON.stringify(this.generateFrameworkDescription(data), null, 2),
        { 'user.name': data.userName, permission: '644', overwrite: 'true' }
      ),
      Promise.each([ 'log', 'tmp', 'finished' ], x => {
        return hdfs.createFolder(
          `/Container/${data.userName}/${name}/` + x,
          { 'user.name': data.userName, permission: '755' }
        );
      }),
      Promise.each([ ...Array(data.taskRoles.length).keys() ], x => {
        return hdfs.createFile(
          `/Container/${data.userName}/${name}/YarnContainerScripts/${x}.sh`,
          this.generateYarnContainerScript(data, x),
          { 'user.name': data.userName, permission: '644', overwrite: 'true' }
        );
      }),
      Promise.each([ ...Array(data.taskRoles.length).keys() ], x => {
        return hdfs.createFile(
          `/Container/${data.userName}/${name}/DockerContainerScripts/${x}.sh`,
          this.generateDockerContainerScript(data, x),
          { 'user.name': data.userName, permission: '644', overwrite: 'true' }
        );
      }),
    ];
    if (!data.originalData.outputDir) {
      hdfsTasks.push(
        hdfs.createFolder(
          `/Output/${data.userName}/${name}`,
          { 'user.name': data.userName, permission: '755' }
        )
      );
    }

    await Promise.all(hdfsTasks);
  }

  generateYarnContainerScript(data, idx) {
    const launcherConfig = this.launcherConfig;
    const jobInfo = {
      idx,
      hdfsUri: launcherConfig.hdfsUri,
      taskData: data.taskRoles[idx],
      jobData: data,
      cluster_id: this.config.clusterId,
    };

    const debugClusters = this.config.virtualDebugClusters;

    if (debugClusters.indexOf(data.virtualCluster) > -1) {
      jobInfo.debug = 1;
    }
    const yarnContainerScript = mustache.render(
      yarnContainerScriptTemplate, jobInfo);
    return yarnContainerScript;
  }

  generateDockerContainerScript(data, idx) {
    const launcherConfig = this.launcherConfig;
    let tasksNumber = 0;
    for (let i = 0; i < data.taskRoles.length; i++) {
      tasksNumber += data.taskRoles[i].taskNumber;
    }

    const jobInfo = {
      idx,
      tasksNumber,
      taskRoleList: data.taskRoles.map(x => x.name).join(','),
      taskRolesNumber: data.taskRoles.length,
      hdfsUri: launcherConfig.hdfsUri,
      taskData: data.taskRoles[idx],
      jobData: data,
    };

    const debugClusters = this.config.virtualDebugClusters;

    if (debugClusters.indexOf(data.virtualCluster) > -1) {
      jobInfo.debug = 1;
    }

    const dockerContainerScript = mustache.render(
      dockerContainerScriptTemplate, jobInfo);
    return dockerContainerScript;
  }
  generateFrameworkDescription(data) {
    const gpuType = data.gpuType || null;
    const fancyRetryPolicy = (data.retryCount !== -2);
    const virtualCluster = (!data.virtualCluster) ? 'default' : data.virtualCluster;
    const frameworkDescription = {
      version: 10,
      user: {
        name: data.userName,
      },
      retryPolicy: {
        maxRetryCount: data.retryCount,
        fancyRetryPolicy,
      },
      taskRoles: {},
      platformSpecificParameters: {
        queue: virtualCluster,
        taskNodeGpuType: gpuType,
        gangAllocation: true,
      },
    };
    for (let i = 0; i < data.taskRoles.length; i++) {
      const portList = {};
      for (let j = 0; j < data.taskRoles[i].portList.length; j++) {
        portList[data.taskRoles[i].portList[j].label] = {
          start: data.taskRoles[i].portList[j].beginAt,
          count: data.taskRoles[i].portList[j].portNumber,
        };
      }
      for (const defaultPortLabel of [ 'http', 'ssh' ]) {
        if (!(defaultPortLabel in portList)) {
          portList[defaultPortLabel] = {
            start: 0,
            count: 1,
          };
        }
      }
      const taskRole = {
        taskNumber: data.taskRoles[i].taskNumber,
        taskService: {
          version: 0,
          entryPoint: `source YarnContainerScripts/${i}.sh`,
          sourceLocations: [ `/Container/${data.userName}/${data.jobName}/YarnContainerScripts` ],
          resource: {
            cpuNumber: data.taskRoles[i].cpuNumber,
            memoryMB: data.taskRoles[i].memoryMB,
            shmMB: data.taskRoles[i].shmMB,
            gpuNumber: data.taskRoles[i].gpuNumber,
            portDefinitions: portList,
            diskType: 0,
            diskMB: 0,
          },
        },
        applicationCompletionPolicy: {
          minFailedTaskCount: data.taskRoles[i].minFailedTaskCount,
          minSucceededTaskCount: data.taskRoles[i].minSucceededTaskCount,
        },
      };
      frameworkDescription.taskRoles[data.taskRoles[i].name] = taskRole;
    }
    return frameworkDescription;
  }

  convertJobState(frameworkState, exitCode) {
    let jobState = '';
    switch (frameworkState) {
      case 'FRAMEWORK_WAITING':
      case 'APPLICATION_CREATED':
      case 'APPLICATION_LAUNCHED':
      case 'APPLICATION_WAITING':
        jobState = 'WAITING';
        break;
      case 'APPLICATION_RUNNING':
      case 'APPLICATION_RETRIEVING_DIAGNOSTICS':
      case 'APPLICATION_COMPLETED':
        jobState = 'RUNNING';
        break;
      case 'FRAMEWORK_COMPLETED':
        if (typeof exitCode !== 'undefined' && parseInt(exitCode) === 0) {
          jobState = 'SUCCEEDED';
        } else if (typeof exitCode !== 'undefined' && parseInt(exitCode) === 214) {
          jobState = 'STOPPED';
        } else {
          jobState = 'FAILED';
        }
        break;
      default:
        jobState = 'UNKNOWN';
    }
    return jobState;
  }

  async generateJobDetail(framework) {
    const jobDetail = {
      jobStatus: {},
      taskRoles: {},
    };
    const frameworkStatus = framework.aggregatedFrameworkStatus.frameworkStatus;
    if (frameworkStatus) {
      const jobState = this.convertJobState(
        frameworkStatus.frameworkState,
        frameworkStatus.applicationExitCode);
      let jobRetryCount = 0;
      const jobRetryCountInfo = frameworkStatus.frameworkRetryPolicyState;
      jobRetryCount =
              jobRetryCountInfo.succeededRetriedCount +
              jobRetryCountInfo.transientNormalRetriedCount +
              jobRetryCountInfo.transientConflictRetriedCount +
              jobRetryCountInfo.nonTransientRetriedCount +
              jobRetryCountInfo.unKnownRetriedCount;
      jobDetail.jobStatus = {
        name: framework.name,
        username: 'unknown',
        state: jobState,
        subState: frameworkStatus.frameworkState,
        executionType: framework.summarizedFrameworkInfo.executionType,
        retries: jobRetryCount,
        createdTime: frameworkStatus.frameworkCreatedTimestamp,
        completedTime: frameworkStatus.frameworkCompletedTimestamp,
        appId: frameworkStatus.applicationId,
        appProgress: frameworkStatus.applicationProgress,
        appTrackingUrl: frameworkStatus.applicationTrackingUrl,
        appLaunchedTime: frameworkStatus.applicationLaunchedTimestamp,
        appCompletedTime: frameworkStatus.applicationCompletedTimestamp,
        appExitCode: frameworkStatus.applicationExitCode,
        appExitDiagnostics: frameworkStatus.applicationExitDiagnostics,
        appExitType: frameworkStatus.applicationExitType,
      };
    }
    const frameworkRequest = framework.aggregatedFrameworkRequest.frameworkRequest;
    if (frameworkRequest.frameworkDescriptor) {
      jobDetail.jobStatus.username = frameworkRequest.frameworkDescriptor.user.name;
    }
    const frameworkInfo = framework.summarizedFrameworkInfo;
    if (frameworkInfo) {
      jobDetail.jobStatus.virtualCluster = frameworkInfo.queue;
    }
    const taskRoleStatuses = framework.aggregatedFrameworkStatus.aggregatedTaskRoleStatuses;
    if (taskRoleStatuses) {
      let ipOfContainer;
      for (const taskRole of Object.keys(taskRoleStatuses)) {
        jobDetail.taskRoles[taskRole] = {
          taskRoleStatus: { name: taskRole },
          taskStatuses: [],
        };
        for (const task of taskRoleStatuses[taskRole].taskStatuses.taskStatusArray) {
          const containerPorts = {};
          if (task.containerPorts) {
            for (const portStr of task.containerPorts.split(';')) {
              if (portStr.length > 0) {
                const port = portStr.split(':');
                containerPorts[port[0]] = port[1];
              }
            }
          }
          const ipNatdb = ipmaptable.getNatConfig(this.config.natFile);
          if (ipNatdb !== null && ipmaptable.getIntraIp(ipNatdb, task.containerIp)) {
            ipOfContainer = ipmaptable.getExtraIp(ipNatdb, task.containerIp);
          } else {
            ipOfContainer = task.containerIp;
          }
          jobDetail.taskRoles[taskRole].taskStatuses.push({
            taskIndex: task.taskIndex,
            containerId: task.containerId,
            containerIp: ipOfContainer,
            containerPorts,
            containerGpus: task.containerGpus,
            containerLog: task.containerLogHttpAddress,
          });
        }
      }
    }
    return jobDetail;
  }

  async getJobList(query) {
    const launcherConfig = this.launcherConfig;
    let reqPath = launcherConfig.frameworksPath();
    if (query.username) {
      reqPath = `${reqPath}?UserName=${query.username}`;
    }

    const resJson = await this._invokeService(reqPath);

    const jobList = resJson.summarizedFrameworkInfos.map(frameworkInfo => {
      let retries = 0;
      [ 'succeededRetriedCount', 'transientNormalRetriedCount', 'transientConflictRetriedCount',
        'nonTransientRetriedCount', 'unKnownRetriedCount' ].forEach(retry => {
        retries += frameworkInfo.frameworkRetryPolicyState[retry];
      });

      return {
        userId: frameworkInfo.userName,
        name: frameworkInfo.frameworkName,
        username: undefined,
        state: this.convertJobState(frameworkInfo.frameworkState, frameworkInfo.applicationExitCode),
        subState: frameworkInfo.frameworkState,
        executionType: frameworkInfo.executionType,
        retries,
        createdTime: frameworkInfo.firstRequestTimestamp || new Date(2018, 1, 1).getTime(),
        completedTime: frameworkInfo.frameworkCompletedTimestamp,
        appExitCode: frameworkInfo.applicationExitCode,
        virtualCluster: frameworkInfo.queue,
      };
    });
    jobList.sort((a, b) => b.createdTime - a.createdTime);
    return jobList;
  }

  async putJobExecutionType(name, data) {
    const launcherConfig = this.launcherConfig;
    const requestResJson = await this._invokeService(launcherConfig.frameworkRequestPath(name));

    if (data.username !== requestResJson.frameworkDescriptor.user.name && !data.admin) {
      throw new LError(ECode.OPERATION_FORBIDDEN, `User ${data.username} is not allowed to execute job ${name}.`);
    }

    const res = await this._invokeService(launcherConfig.frameworkExecutionTypePath(name), {
      method: 'PUT',
      data: { executionType: data.value },
    });
    return res;
  }

  async getJobConfig(userName, jobName) {
    const launcherConfig = this.launcherConfig;
    const hdfs = new Hdfs(launcherConfig.webhdfsUri);
    const result = await hdfs.readFile(
      `/Container/${userName}/${jobName}/JobConfig.json`,
      null);
    return JSON.parse(result.content);
  }

  async getJobSshInfo(userName, jobName, applicationId) {
    const clusterId = this.config.clusterId;
    const folderPathPrefix = `/gpai/${clusterId}/${userName}/${jobName}/ssh/${applicationId}`;
    const readdir = Promise.promisify(fs.readdir);
    const result = await readdir(folderPathPrefix);
    const { stdout } = await execa.shell(`cat ${folderPathPrefix}/.ssh/${applicationId}`);

    const sshkey = stdout;
    const sshInfo = {
      containers: [],
      keyPair: {
        privateKey: sshkey,
        privateKeyFileName: applicationId,
      },
    };
    let extraSshIp,
      extraSshPort;
    for (const x of result) {
      const pattern = /^container_(.*)-(.*)-(.*)$/g;
      const arr = pattern.exec(x);
      if (arr !== null) {
        const sshIp = arr[2].toString();
        const sshPort = arr[3].toString();
        const ipNatdb = ipmaptable.getNatConfig(this.config.natFile);
        this.logger.info(ipmaptable.getIntraIp(ipNatdb, sshIp));
        if (ipmaptable.getIntraIp(ipNatdb, sshIp)) {
          extraSshIp = ipmaptable.getExtraIp(ipNatdb, sshIp);
          extraSshPort = ipmaptable.getExtraPort(ipNatdb, sshIp, sshPort);
        } else {
          extraSshIp = arr[2];
          extraSshPort = arr[3];
        }
        sshInfo.containers.push({
          id: 'container_' + arr[1],
          sshIp: extraSshIp,
          sshPort: extraSshPort,
        });
      }
    }
    return sshInfo;
  }

  async deleteJob(name, data) {
    const launcherConfig = this.launcherConfig;
    const job = await this._invokeService(launcherConfig.frameworkRequestPath(name));
    if (data.username !== job.frameworkDescriptor.user.name && !data.admin) {
      throw new LError(ECode.OPERATION_FORBIDDEN, `User ${data.username} is not allowed to remove job ${name}.`);
    }
    await this._invokeService(launcherConfig.frameworkPath(name), { method: 'delete' });
  }
}

module.exports = TokenService;

File: taskset/pkg/pipeline/controllers/v1/job.go
package v1

// MIT License
//
// Copyright (c) PCL. All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE
//
import (
	"fmt"
	"net/http"
	v1 "scheduler/pkg/pipeline/apis/http/v1"
	"scheduler/pkg/pipeline/apis/module"
	"scheduler/pkg/pipeline/app"
	"scheduler/pkg/pipeline/constants/header"
	"scheduler/pkg/pipeline/constants/jobstate"
	status "scheduler/pkg/pipeline/constants/statusphrase"
	formatTaskSet "scheduler/pkg/pipeline/utils/taskset"
	typeTaskset "scheduler/pkg/crd/apis/taskset/v1alpha1"
	"strconv"
	"strings"
	"time"
	"encoding/json"
	"github.com/gin-gonic/gin"
	jsoniter "github.com/json-iterator/go"
	uuid "github.com/satori/go.uuid"
)

func SubmitJob(app *app.App, c *gin.Context) error {

	var param v1.SubmitJobParam

	if err := c.ShouldBindJSON(&param); err != nil {
		return err
	}

	if "" == param.UserID || "" == param.JobKind || "" == param.JobName {
		c.JSON(http.StatusOK, gin.H{
			"code": status.LACK_OF_PARAM,
			"msg":  "Missing parameter 'userID' 、'jobKind' or 'jobName'",
		})
		return nil
	}

	var jobID string

	job, err := jsoniter.Marshal(param.Job)

	if err != nil {
		return err
	}

	kind := jsoniter.Get(job, "kind").ToString()

	if "" == kind {
		c.JSON(http.StatusOK, gin.H{
			"code": status.WRONG_PARAM,
			"msg":  "job.kind can't be empty",
		})
		return nil
	}

	if strings.ToLower(kind) == "taskset" {
		//if job is taskset ,and name is provided
		jobID = jsoniter.Get(job, "metadata").Get("name").ToString()
	}

	if "" == jobID {
		jobID = jsoniter.Get(job, "jobID").ToString()
	}
	

	if "" == jobID {
		jobID = strings.ReplaceAll(fmt.Sprintf("%s", uuid.NewV4()), "-", "0")
	}

	if nil == param.Header {
		param.Header = map[string]interface{}{}
	}

	param.Header[header.DefaultHeaderJobID] = jobID
	param.Header[header.DefaultHeaderUserID] = param.UserID
	param.Header[header.DefaultHeaderJobKind] = param.JobKind
	param.Header[header.DefaultHeaderJobName] = param.JobName
	param.Header[header.DefaultHeaderCluster] = param.Cluster

	header, err := jsoniter.Marshal(param.Header)

	if err != nil {
		return err
	}

	jobNamespace := param.JobNamespace
	if jobNamespace == "" {
		jobNamespace = "default"
	} else {
		jobNamespace = strings.ToLower(jobNamespace)
	}

	jobRecord := &module.JobRecord {
		ID: jobID,
		UserID: param.UserID,
		Name: param.JobName,
		Namespace: jobNamespace,
		Kind: param.JobKind,
		Header: string(header),
		Cluster: param.Cluster,
	}
	err = app.Services().Core().SubmitJob(jobRecord, job)

	if err != nil {
		return err
	}

	c.JSON(http.StatusOK, gin.H{
		"code": status.OPERATION_SUCCEEDED,
		"msg":  "sucess",
		"payload": gin.H{
			"jobID": jobID,
		},
	})

	return nil
}

func StopJob(app *app.App, ctx *gin.Context) error {

	var param v1.UpdateJobParam

	if err := ctx.ShouldBindJSON(&param); err != nil {
		return err
	}

	if "" == param.JobID || "" == param.Reason {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.LACK_OF_PARAM,
			"msg":  "Missing parameter 'jobID' or 'reason'",
		})
		return nil
	}

	record, err := app.Services().Job().GetJobRecord(param.JobID)

	if err != nil {
		return err
	}

	if nil == record {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.OPEERATION_TARGET_NOT_FOUND,
			"msg":  "Job Not Found",
		})
		return nil
	}

	if record.State == jobstate.SUCCEEDED ||
		record.State == jobstate.FAILED {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.UNSUPPORTED_OPERATION,
			"msg":  fmt.Sprintf("Can't stop a completed job,job is %s", record.State),
		})

		return nil
	}

	if record.State != jobstate.STOPPED {
		err = app.Services().Core().StopJob(param.JobID, record.Namespace, param.Reason)

		if nil != err {
			return err
		}
	}

	ctx.JSON(http.StatusOK, gin.H{
		"code": status.OPERATION_SUCCEEDED,
		"msg":  "sucess",
	})

	return nil
}

func ResumeJob(app *app.App, ctx *gin.Context) error {

	var param v1.UpdateJobParam

	if err := ctx.ShouldBindJSON(&param); err != nil {

		return err
	}

	if "" == param.JobID || "" == param.Reason {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.LACK_OF_PARAM,
			"msg":  "Missing parameter 'jobID' or 'reason'",
		})
		return nil
	}

	cursor, err := app.Services().Job().GetJobCursor(param.JobID)

	if nil != err {
		return err
	}

	if nil == cursor {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.OPEERATION_TARGET_NOT_FOUND,
			"msg": fmt.Sprintf(
				"Can't resume the job:%s,Maybe it does not exist or has already been submited to kubernetes",
				param.JobID),
		})
		return nil
	}

	if cursor.Submited == true {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.UNSUPPORTED_OPERATION,
			"msg":  fmt.Sprintf("Can't resume the job:%s,it has already been submited to kubernetes", param.JobID),
		})
		return nil
	}

	record, err := app.Services().Job().GetJobRecord(param.JobID)

	if nil != err {
		return err
	}

	if nil == record {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.INTERNAL_ERROR,
			"msg":  fmt.Sprintf("The data of job (%s) is lost ", param.JobID),
		})
		return nil
	}

	if record.State != jobstate.SUSPENDED {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.UNSUPPORTED_OPERATION,
			"msg":  fmt.Sprintf("Can't resume job with state: %s", record.State),
		})
		return nil
	}

	var header string = "{}"

	if nil != record && "" != record.Header {
		header = record.Header
	}

	cursor.Header = header

	err = app.Services().Core().ResumeJob(cursor, param.Reason)

	if nil != err {
		return err
	}

	ctx.JSON(http.StatusOK, gin.H{
		"code": status.OPERATION_SUCCEEDED,
		"msg":  "sucess",
	})
	return nil
}

func defaultDetail(record *module.JobRecord) *v1.JobStatusDetail {
	state := jobstate.PREPARING

	if record.State == jobstate.STOPPED ||
		jobstate.SUSPENDED == record.State ||
		jobstate.FAILED == record.State {
		state = record.State
	}

	return &v1.JobStatusDetail{
		Version: "v1",
		Job: &v1.JobSummary{
			ID:              record.ID,
			Name:            record.Name,
			Type:            record.Kind,
			UserID:          record.UserID,
			State:           state,
			ExitDiagnostics: record.StateSummary,
		},
		Cluster: &v1.ClusterInfo{ Identity: record.Cluster },
		PlatformSpecificInfo: &v1.PlatformSpecificInfo{
			Platform: "k8s",
		},
	}
}

func FormatWebStateSummary(record *module.JobRecord){

	var messages []typeTaskset.TaskMessage = []typeTaskset.TaskMessage{}

	err := json.Unmarshal([]byte(record.StateSummary), &messages)

	if nil != err {
		fmt.Println(err)
		return
	}

	var taskMessage = ""

	for i := 0; i < len(messages); i++ {

		var message = messages[i]
		taskMessage += "[task] state: " + message.State + "\n"
		taskMessage += "[task] stateMessage: " + message.StateMessage + "\n"
		for j := 0; j < len(message.Roles); j++ {
			var role = message.Roles[j]
			var name = role.Name
			var phase = role.Phase
			var phaseMessage  = role.PhaseMessage
			taskMessage += "[role " + name + "] phase: " + phase + "\n"
			taskMessage += "[role " + name + "] phaseMessage: " + phaseMessage + "\n"
			for k := 0; k < len(role.Replicas); k++ {
				var replica = role.Replicas[k]
				var name = replica.Name
				var phase = replica.Phase
				var phaseMessage  = replica.PhaseMessage
				taskMessage += "[replica " + name + "] phase: " + phase + "\n"
				taskMessage += "[replica " + name + "] phaseMessage: " + phaseMessage + "\n"
			}
		}

		taskMessage += "\n"
	}
	record.StateSummary = taskMessage
}


func GetJobDetail(app *app.App, ctx *gin.Context) error {

	jobID := ctx.Param("job")

	record, err := app.Services().Job().GetJobRecord(jobID)

	if err != nil {
		return err
	}

	if nil == record {
		ctx.JSON(http.StatusNotFound, gin.H{
			"code": status.OPEERATION_TARGET_NOT_FOUND,
			"msg":  fmt.Sprintf("Not Found"),
		})
		return nil
	}

	state := record.State

	FormatWebStateSummary(record)

	if state == jobstate.FAILED || state == jobstate.STOPPED || state == jobstate.SUCCEEDED {
		if "" == record.Detail || "{}" == record.Detail {
			ctx.JSON(http.StatusOK, gin.H{
				"code":    status.OPERATION_SUCCEEDED,
				"msg":     "success",
				"payload": defaultDetail(record),
			})
			return nil
		}

		var detail v1.JobStatusDetail
		json.Unmarshal([]byte(record.Detail), &detail)
		detail.Job.ExitDiagnostics = record.StateSummary
		bytes, _ := json.Marshal(detail)
		record.Detail = string(bytes)

		rsp := `{
			"code":"` + status.OPERATION_SUCCEEDED + `",` + `
			"msg":"success",
			"payload":` + record.Detail + `
		}`
		ctx.Data(http.StatusOK, "application/json", []byte(rsp))
		return nil
	}

	var namespace string = record.Namespace

	if "" == namespace {
		namespace = "default"
	}

	taskset, err := app.Services().Kubernetes().GetJob(namespace, jobID)

	if nil != err {
		return err
	}

	var detailStatus *v1.JobStatusDetail

	detailStatus = formatTaskSet.Format(record.Name, record.Kind, record.UserID, record.Cluster, record.StateSummary, taskset)

	if nil == detailStatus {
		detailStatus = defaultDetail(record)
	}

	ctx.JSON(http.StatusOK, gin.H{
		"code":    status.OPERATION_SUCCEEDED,
		"msg":     "success",
		"payload": detailStatus,
	})
	return nil
}

func GetJobConfig(app *app.App, ctx *gin.Context) error {

	jobID := ctx.Param("job")

	record, err := app.Services().Job().GetJobRecord(jobID)

	if nil != err {
		return err
	}

	if nil == record {
		ctx.JSON(http.StatusNotFound, gin.H{
			"code": status.OPEERATION_TARGET_NOT_FOUND,
			"msg":  fmt.Sprintf("Not Found"),
		})
		return nil
	}

	rsp := `{
		"code":"` + status.OPERATION_SUCCEEDED + `",` + `
		"msg":"success",
		"payload":` + record.Config + `
	}`

	ctx.Data(http.StatusOK, "application/json", []byte(rsp))

	return nil
}

func GetJobTaskSet(app *app.App, ctx *gin.Context) error {

	jobID := ctx.Param("job")

	cursor, err := app.Services().Job().GetJobCursor(jobID)

	if nil != err {
		return err
	}

	if nil != cursor && "" != cursor.Job {
		rsp := `{
			"code":"` + status.OPERATION_SUCCEEDED + `",` + `
			"msg":"success",
			"payload":` + cursor.Job + `
		}`
		ctx.Data(http.StatusOK, "application/json", []byte(rsp))
		return nil
	}

	ctx.JSON(http.StatusNotFound, gin.H{
		"code": status.OPEERATION_TARGET_NOT_FOUND,
		"msg":  "Not Found",
	})
	return nil
}

func parseQuery(ctx *gin.Context) *v1.QueryParams {
	query := &v1.QueryParams{}

	if ctx.Query("userID") != "" {
		query.UserID = ctx.Query("userID")
	}
	if ctx.Query("cluster") != "" {
		query.Cluster = ctx.Query("cluster")
	}

	if ctx.Query("kind") != "" {
		query.JobKind = ctx.Query("kind")
	}

	if ctx.Query("name") != "" {
		query.JobName = ctx.Query("name")
	}

	if ctx.Query("state") != "" {
		query.State = ctx.Query("state")
	}

	if ctx.Query("order") != "" {
		query.Order = ctx.Query("order")
	}

	if ctx.Query("pageNumber") != "" {
		pageNumber, err := strconv.ParseInt(ctx.Query("pageNumber"), 10, 64)
		if err == nil && pageNumber > 0 {
			query.PageNumber = pageNumber
		}
	}

	if ctx.Query("pageSize") != "" {
		pageSize, err := strconv.ParseInt(ctx.Query("pageSize"), 10, 64)
		if err == nil && pageSize > 0 {
			query.PageSize = pageSize
		}
	}

	if ctx.Query("startAt") != "" {
		second, err := strconv.ParseInt(ctx.Query("startAt"), 10, 64)
		if err == nil && second > 0 {
			startAt := time.Unix(second, 0)
			query.StartTime = &startAt
		}
	}

	if ctx.Query("endAt") != "" {
		second, err := strconv.ParseInt(ctx.Query("endAt"), 10, 64)
		if err == nil && second > 0 {
			endAt := time.Unix(second, 0)
			query.EndTime = &endAt
		}
	}
	return query
}

func GetJobList(app *app.App, ctx *gin.Context) error {

	query := parseQuery(ctx)

	if 0 == query.PageNumber || 0 == query.PageSize {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.WRONG_PARAM,
			"msg":  "Query Parameter 'pageSize' and 'pageNumber' can't be zero",
		})
		return nil
	}

	if query.PageSize != 10 && query.PageSize != 20 && query.PageSize != 50 {
		ctx.JSON(http.StatusOK, gin.H{
			"code": status.WRONG_PARAM,
			"msg":  "Query Parameter 'pageSize' can only be 10、20　or 50",
		})
		return nil
	}

	list, err := app.Services().Job().SearchJobs(query)

	if err != nil {
		return err
	}

	ctx.JSON(http.StatusOK, gin.H{
		"code": status.OPERATION_SUCCEEDED,
		"msg":  "success",
		"payload": gin.H{
			"jobs": list,
		},
	})

	return nil
}

func GetJobCount(app *app.App, ctx *gin.Context) error {
	query := parseQuery(ctx)

	amount, err := app.Services().Job().GetJobAmount(query)

	if err != nil {
		return err
	}

	ctx.JSON(http.StatusOK, gin.H{
		"code": status.OPERATION_SUCCEEDED,
		"msg":  "success",
		"payload": gin.H{
			"count": amount,
		},
	})
	return nil
}
